### PR TITLE
Drop python 2

### LINKED ISSUE #3562: Drop python 2 support

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #3562

### CODE REVIEW

File: setup.json [hunk 1]
```diff
@@ -84,15 +80,12 @@
...
+      "sphinxcontrib-contentui==0.2.4",
+      "sphinxcontrib-details-directive==0.1.0"
     ],
     "atomic_tools": [
-      "PyCifRW==4.2.1; python_version < '3'",
-      "PyCifRW==4.4.1; python_version >= '3'",
+      "PyCifRW==4.4.1",
       "ase==3.17.0",
       "monty==2.0.4",
       "pymatgen<=2018.12.12",
```
Comment (greschd) [inline setup.json]: I think for ``pymatgen`` and ``monty`` the main reason to use the pre-2019 version was also py2 compatibility.

File: aiida/sphinxext/__init__.py [hunk 2]
```diff
@@ -11,10 +11,6 @@
 Defines reStructuredText directives to simplify documenting AiiDA and its plugins.
 """
 
-from __future__ import division
-from __future__ import print_function
-from __future__ import absolute_import
-
 from sphinx.errors import ExtensionError
```
Comment (greschd) [inline aiida/sphinxext/__init__.py]: The try / except just below this is also a py2 workaround (for some reason I can't comment there :upside_down_face:)

File: aiida/transports/transport.py [hunk 3]
```diff
@@ -41,8 +35,7 @@ def validate_positive_number(ctx, param, value):  # pylint: disable=unused-argum
         raise BadParameter('{} is not a valid positive number'.format(value))
 
 
-@six.add_metaclass(ABCMeta)
-class Transport(object):
```
Comment (greschd) [inline aiida/transports/transport.py]: Same comment about the metaclass as above.

File: aiida/backends/general/abstractqueries.py [hunk 4]
```diff
@@ -7,16 +7,11 @@
...
-from __future__ import print_function
-from __future__ import absolute_import
 
 from abc import ABCMeta
-import six
 
 
-@six.add_metaclass(ABCMeta)
-class AbstractQueryManager(object):
+class AbstractQueryManager:
```
Comment (greschd) [inline aiida/backends/general/abstractqueries.py]: Should the metaclass actually be removed, or just put with new syntax? Could also inherit from ``ABC`` instead.

File: aiida/backends/testimplbase.py [hunk 5]
```diff
@@ -7,19 +7,14 @@
...
-from __future__ import absolute_import
 from abc import ABCMeta, abstractmethod
 
-import six
 
 from aiida import orm
 from aiida.common import exceptions
 
 
-@six.add_metaclass(ABCMeta)
```
Comment (greschd) [inline aiida/backends/testimplbase.py]: Same as above: should the metaclass be removed?

File: aiida/backends/tests/cmdline/commands/test_data.py [hunk 6]
```diff
@@ -19,7 +16,6 @@
 import numpy as np
 import subprocess as sp
 
-from six.moves import cStringIO as StringIO
```
Comment (greschd) [inline aiida/backends/tests/cmdline/commands/test_data.py]: Is there a ``from io import StringIO`` missing? Or maybe just put it as ``io.StringIO`` below?

File: aiida/backends/tests/test_nodes.py [hunk 7]
```diff
@@ -1593,16 +1585,8 @@ def _check_existence(self, uuids_check_existence, uuids_check_deleted):
...
-            import sys
-            six.reraise(
-                type(exc),
-                exc.__class__((
-                    str(exc) + '\nCurrent UUID being processed: {}' + '\nFull uuids_check_existence: {}' +
-                    '\nFull uuids_check_deleted: {}'
-                ).format(uuid, uuids_check_existence, uuids_check_deleted)),
-                sys.exc_info()[2]
-            )
+        except Exception:  # pylint: disable=broad-except
```
Comment (greschd) [inline aiida/backends/tests/test_nodes.py]: This try/except doesn't do anything, right?

File: aiida/manage/backup/backup_base.py [hunk 8]
```diff
@@ -8,27 +8,21 @@
...
 from dateutil.parser import parse
 from pytz import timezone as ptimezone
 
 from aiida.common import json
 from aiida.common import timezone as dtimezone
 
 
-@six.add_metaclass(ABCMeta)  # pylint: disable=useless-object-inheritance
-class AbstractBackup(object):
+class AbstractBackup:
```
Comment (greschd) [inline aiida/manage/backup/backup_base.py]: Same here about the metaclass.

File: aiida/orm/implementation/backends.py [hunk 9]
```diff
@@ -8,21 +8,16 @@
...
-import six
 
 __all__ = ('Backend', 'BackendEntity', 'BackendCollection', 'EntityType')
 
 EntityType = typing.TypeVar('EntityType')  # pylint: disable=invalid-name
 
 
-@six.add_metaclass(abc.ABCMeta)
-class Backend(object):
+class Backend:
```
Comment (greschd) [inline aiida/orm/implementation/backends.py]: Same here, is the metaclass needed?

File: aiida/orm/implementation/backends.py [hunk 10]
```diff
@@ -121,8 +116,7 @@ def transaction(self):
         """
 
 
-@six.add_metaclass(abc.ABCMeta)
-class BackendEntity(object):
+class BackendEntity:
```
Comment (greschd) [inline aiida/orm/implementation/backends.py]: Same here, is the metaclass needed?

File: aiida/parsers/parser.py [hunk 11]
```diff
@@ -25,8 +20,7 @@
 __all__ = ('Parser',)
 
 
-@six.add_metaclass(ABCMeta)
```
Comment (greschd) [inline aiida/parsers/parser.py]: metaclass?

File: aiida/tools/dbimporters/plugins/materialsproject.py [hunk 12]
```diff
@@ -45,13 +41,11 @@ def setup_db(self, **kwargs):
...
                 api_key = os.environ['PMG_MAPI_KEY']
-            except KeyError as exc:
-                raise_from(
-                    KeyError(
-                        'API key not supplied and PMG_MAPI_KEY environment '
-                        'variable not set. Either pass it when initializing the class, '
-                        'or set the environment variable PMG_MAPI_KEY to your API key.'
-                    ), exc
+            except KeyError:
+                KeyError(
```
Comment (greschd) [inline aiida/tools/dbimporters/plugins/materialsproject.py]: This should probably be 
```
except KeyError as exc:
    raise KeyError(...) from exc
```

That way you get the proper "error chaining" (``The above exception was the direct cause of the following exception:`` instead of ``During handling of the above exception, another exception occurred:``).

File: setup.json (hunk 1 above)
Comment (greschd) [reply setup.json]: Although we can of course update these later as well.

Comment (sphuber) [reply setup.json]: I was leaving all those dependencies to be dealt with in #3547 separately. We intend to relax the pinning as much as possible but that will require quite some testing. For now I just removed those that are explicitly for python<3.5 since that is no longer supported

File: aiida/transports/transport.py (hunk 3 above)
Comment (sphuber) [reply aiida/transports/transport.py]: Good point. I put it for the `Node` class which used the `AbstractNodeMeta` but not sure if it is actually _needed_ for any of the others. Do you know when using ABC as metaclass is crucial as opposed to just making a class with abstract methods? Will the latter fail if class is not explicitly marked as abstract as well?

File: aiida/backends/tests/cmdline/commands/test_data.py (hunk 6 above)
Comment (sphuber) [reply aiida/backends/tests/cmdline/commands/test_data.py]: no, but there is `import io` at the top and just used `io.StringIO`

File: aiida/transports/transport.py (hunk 3 above)
Comment (sphuber) [reply aiida/transports/transport.py]: I think I understand now why nothing complained and everything still works when I just removed it. Subclassing from `ABC` is the same as setting the meta to `ABCMeta`. It is just a shortcut. Having `ABCMeta` as meta class is _necessary_ if you want to mark methods as abstract with `@abstractmethod`. Looking at the code where I removed it though, which is mostly very old code, it seems that the original writers never used the `@abstractmethod` decorators, but just raised `NotImplementedError`. This is not statically checked at launch time, unlike the decorators. Instantiating a sub class with an abstract method will raise, but if you don't override the method of the base class, no-one will notice until you call it.

In conclusion, I think these should for the most part be abstract classes but then their methods should be appropriately made abstract as well. The question is now, do I do this in this PR? Or shall I for now just make them subclass `ABC` directly and leave the fixing of the methods for another time?

File: aiida/backends/tests/test_nodes.py (hunk 7 above)
Comment (sphuber) [reply aiida/backends/tests/test_nodes.py]: Correct, removed

File: aiida/transports/transport.py (hunk 3 above)
Comment (greschd) [reply aiida/transports/transport.py]: I see, that makes sense.

For the sake of making it a bit easier to track the change, maybe it's best to just subclass ``ABC`` now, and then make a separate PR with the abstractmethods.

File: aiida/tools/dbimporters/plugins/materialsproject.py (hunk 12 above)
Comment (sphuber) [reply aiida/tools/dbimporters/plugins/materialsproject.py]: Done

File: aiida/transports/transport.py (hunk 3 above)
Comment (sphuber) [reply aiida/transports/transport.py]: Done

File: aiida/backends/tests/cmdline/commands/test_data.py (hunk 6 above)
Comment (greschd) [reply aiida/backends/tests/cmdline/commands/test_data.py]: Isn't there a bare ``StringIO`` [here](https://github.com/sphuber/aiida_core/blob/fix_3562_drop_python_2/aiida/backends/tests/cmdline/commands/test_data.py#L33)? I _think_ I'm looking at the right commit 😅

Comment (sphuber) [reply aiida/backends/tests/cmdline/commands/test_data.py]: yep you were right, but the function that it was in was never used :sweat_smile: the calls were commented out, so I will just remove the function

**Comment by greschd** [comment]
I think the only thing left to do now is telling Jenkins to use py3 as well.

**Comment by sphuber** [comment]
Yes, but I also wanted to merge some final things that should go into `v1.0.1` that I want to release soon. Then we can merge this in `develop` which is then py3 only and this will become `v1.1.0`. We are about to accept the AEP that defines that version as the one where we drop py2 support.

**Comment by greschd** [comment]
@sphuber I've just stumbled across a *very important* little detail: In order for the "graceful fallback" on ``pip install`` to work for older versions, we need to *first* release another version that also specifies ``python_requires``. No idea why, but 

So that means, for the ``1.0.1`` release we should add ``python_requires=">=2.7"`` in ``setup.json``. 

No idea why that's the case, but just tested this with a different package. It's described [here (last paragraph)](https://packaging.python.org/guides/dropping-older-python-versions/#dropping-a-python-release), although you could easily miss it.

**Comment by sphuber** [comment]
Good catch! I will open a PR now

**Comment by sphuber** [comment]
@greschd if you'd like, I added three more commits on top since last time you reviewed. If you have the time you can just look at those commits.

File: aiida/common/hashing.py [hunk 13]
```diff
@@ -160,19 +160,13 @@ def _single_digest(obj_type, obj_bytes=b''):
 
 @_make_hash.register(bytes)
 def _(bytes_obj, **kwargs):
-    """
-    Hash arbitrary binary strings (str in Python 2, bytes in Python 3).
-    For compat reason between Python 2 and 3, this gets the same hash-type
-    as for unicode in Python 2, resp. str in Python 3."""
+    """Hash arbitrary byte strings."""
     return [_single_digest('str', bytes_obj)]
```
Comment (greschd) [inline aiida/common/hashing.py]: Hmm, I'm not sure if we might want to change the "type specifier" here to be bytes - specific, now that py2 / py3 compatibility is no longer a concern.
